pkg/sbom/spdx/unmarshal.go

```diff
@@ -0,0 +1,238 @@
+		return xerrors.Errorf("failed to load spdx json: %w", err)
+	}
+	err = s.unmarshal(spdxDocument)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal spdx: %w", err)
+	}
+	return nil
+}
+
+func (s *SPDX) unmarshal(spdxDocument *spdx.Document2_2) error {
+	s.relationships = relationshipMap(spdxDocument.Relationships)
+	s.packages = spdxDocument.Packages
+
+	for pkgID := range s.relationships {
+		pkg := s.packages[pkgID]
+		switch {
+		case strings.HasPrefix(string(pkg.PackageSPDXIdentifier), ElementOperatingSystem):
+			s.SBOM.OS = parseOS(pkg)
+			pkgs, err := s.parsePkgs(pkg.PackageSPDXIdentifier)
+			if err != nil {
+				return xerrors.Errorf("failed to parse os packages: %w", err)
+			}
+			if len(pkgs) != 0 {
+				s.SBOM.Packages = []ftypes.PackageInfo{{Packages: pkgs}}
+			}
+
+		case strings.HasPrefix(string(pkg.PackageSPDXIdentifier), ElementApplication):
+			app, err := s.parseApplication(pkg)
+			if err != nil {
+				return xerrors.Errorf("failed to parse application: %w", err)
+			}
+			s.SBOM.Applications = append(s.SBOM.Applications, *app)
+		}
+	}
+
+	s.SPDX = spdxDocument
+	return nil
+}
+
+func (s *SPDX) parseApplication(pkg *spdx.Package2_2) (*ftypes.Application, error) {
+	pkgs, err := s.parsePkgs(pkg.PackageSPDXIdentifier)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to parse language packages: %w", err)
+	}
+	app := &ftypes.Application{
+		Type:      pkg.PackageName,
+		FilePath:  pkg.PackageSourceInfo,
+		Libraries: pkgs,
+	}
+	if pkg.PackageName == ftypes.NodePkg || pkg.PackageName == ftypes.PythonPkg ||
+		pkg.PackageName == ftypes.GemSpec || pkg.PackageName == ftypes.Jar {
+		app.FilePath = ""
+	}
+	return app, nil
+
+}
+
+func (s *SPDX) parsePkgs(id spdx.ElementID) ([]ftypes.Package, error) {
+	pkgIDs := s.relationships[id]
+
+	var pkgs []ftypes.Package
+	for _, id := range pkgIDs {
+		spdxPkg := s.packages[id]
+		pkg, err := parsePkg(spdxPkg)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to parse package: %w", err)
+		}
+
+		pkgs = append(pkgs, *pkg)
+	}
+	return pkgs, nil
+}
+
+func parseOS(pkg *spdx.Package2_2) *ftypes.OS {
+	return &ftypes.OS{
+		Family: pkg.PackageName,
+		Name:   pkg.PackageVersion,
+	}
+}
+
+func parsePkg(package2_2 *spdx.Package2_2) (*ftypes.Package, error) {
+	var (
+		pkg *ftypes.Package
+		typ string
+	)
+	for _, ref := range package2_2.PackageExternalReferences {
+		if ref.RefType == RefTypePurl && ref.Category == CategoryPackageManager {
+			packageURL, err := purl.FromString(ref.Locator)
+			if err != nil {
+				return nil, xerrors.Errorf("failed to parse purl from string: %w", err)
+			}
+			pkg = packageURL.Package()
+			pkg.Ref = ref.Locator
+			typ = packageURL.Type
+			break
+		}
+	}
+	if pkg == nil {
+		return nil, errInvalidPackageFormat
+	}
+
+	if package2_2.PackageLicenseDeclared != "NONE" {
+		pkg.Licenses = strings.Split(package2_2.PackageLicenseDeclared, ",")
+	}
+	pkg.Name = package2_2.PackageName
+	pkg.Version = package2_2.PackageVersion
```

Comment:
Do we need to set them though the name and version are already obtained from PURL?